core/src/main/java/org/pentaho/di/connections/utils/ConnectionUriParser.java

```diff
@@ -0,0 +1,229 @@
+/*!
+ * This program is free software; you can redistribute it and/or modify it under the
+ * terms of the GNU Lesser General Public License, version 2.1 as published by the Free Software
+ * Foundation.
+ *
+ * You should have received a copy of the GNU Lesser General Public License along with this
+ * program; if not, you can obtain a copy at http://www.gnu.org/licenses/old-licenses/lgpl-2.1.html
+ * or from the Free Software Foundation, Inc.,
+ * 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * This program is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY;
+ * without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
+ * See the GNU Lesser General Public License for more details.
+ *
+ * Copyright (c) 2024 Hitachi Vantara. All rights reserved.
+ */
+
+package org.pentaho.di.connections.utils;
+
+import java.util.Arrays;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+
+/**
+ * Helper class to extract information from the
+ * {@value org.pentaho.di.connections.vfs.provider.ConnectionFileProvider#SCHEME} URI and more general VFS URIs.
+ * This class handles special characters in the connection name section of the
+ * {@value org.pentaho.di.connections.vfs.provider.ConnectionFileProvider#SCHEME} URI by simply parsing
+ * the connection name as is. This class has limited functionality.
+ * <p/>
+ * Main motivation for this class is not being able to use directly {@link java.net.URI}
+ * due to special characters in the {@value org.pentaho.di.connections.vfs.provider.ConnectionFileProvider#SCHEME} URI.
+ * Future efforts should look into {@link org.apache.commons.vfs2.provider.UriParser#encode(String, char[])}.
+ *
+ * @see #SPECIAL_CHARACTERS_FULL_SET
+ */
+public class ConnectionUriParser {
+
+  /**
+   * Full set of special characters the connection name can be. Only excluding the character '/' which is a file path deliminator.
+   * <p/>
+   * Special characters in this context are characters that generally have to be encoded
+   * otherwise the use of {@link java.net.URI } will throw {@link java.net.URISyntaxException} when during
+   * object instantiation or subsequent method calls.
+   *
+   */
+  public static final String SPECIAL_CHARACTERS_FULL_SET = "!\"#$%&'()*+,-.:;<=>?@[\\] \t\n\r^_`{|}~";
+
+  /**
+   * Full set of alphanumeric characters the connection name can be.
+   */
+  public static final String ALPHANUMERIC_CHARACTERS_FULL_SET =
+      "0123456789ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz";
+
+  /**
+   * Full set of accepted characters the connection name can be.
+   */
+  public static final String ACCEPTED_CHARACTERS_FULL_SET =
+      SPECIAL_CHARACTERS_FULL_SET + ALPHANUMERIC_CHARACTERS_FULL_SET;
```

Comment:
These constants appear to only be used by the test class.
From an API standpoint, don't seem useful, or general enough (e.g. no international characters).
I suggest these are moved to the test class.

Reply:
I would be fine with an update.

kinda re-iterating some points from https://github.com/pentaho/pentaho-kettle/pull/9310#discussion_r1605126415

1. keeping `ConnectionUriParser#SPECIAL_CHARACTERS_FULL_SET` more of less. I think it's important enough to document the special characters we will have in the source code class `ConnectionUriParser.java` and not `ConnectionUriParserTest.java`. Wether documentation includes a combination of static variable, referencing the static variable in javadoc comment, in javadoc referencing static variable or just hand writing the characters in the javadoc or referencing
2. The code supports international characters, so just adding that verbiage in the javadoc. Then for `ConnectionUriParser#ALPHANUMERIC_CHARACTERS_FULL_SET,` removing it from the source code and/or adding to test code. Then `ConnectionUriParser#ACCEPTED_CHARACTERS_FULL_SET` is obsolete by the verbiage update.

Reply:
Addressed in commit: https://github.com/pentaho/pentaho-kettle/commit/57a7b9ed55209fb93c39549b26b39f2a50442921

Reply:
In my opinion, and according to what I think was the result of the team discussion on this, at this moment in time, this code should remain as much as possible agnostic to which are the special characters allowed by connection names. As such, for now, all characters except `/` and `\` should be allowed (PUC is enforcing this, but it seems PDI is not, and is failing in an odd way when trying to create a connection with such a name).

Parsing pvfs URLs with a connection name with any of these characters is met perfectly by the proposed regular expressions, such as: `"^(\\w+)://([^/]+)/?"` without any need for `SPECIAL_CHARACTERS_FULL_SET`.

Depending on what comes to be the decision for the allowed characters, it may be best to write a regexp which lists all allowed characters/classes, or one which excludes certain special characters/classes. My point is, we don't know yet, and so it would be best to leave this out of the public API for now.

Reply:
I'm trying to triage all comments for this PR. I've stated my position and you have stated your position very well.

Is this a item that should prevent merging as ?

If so, we should  go offline and get a technical moderator to make a final decision.

Reply:
Resolving. See https://github.com/pentaho/pentaho-kettle/pull/9310#pullrequestreview-2065857215.